src/jit/lower.cpp

```diff
@@ -4184,8 +4184,17 @@ GenTree* Lowering::LowerVirtualStubCall(GenTreeCall* call)
 // On ARM we must use a proper address in R12(thunk register) without dereferencing.
 // So for the jump we use the default register.
 // TODO: specifying register probably unnecessary for other platforms, too.
-#if !defined(_TARGET_UNIX_) && !defined(_TARGET_ARM_)
+#if !defined(_TARGET_UNIX_) && !defined(_TARGET_ARM_) && !defined(_TARGET_ARM64_)
             indir->gtRegNum = REG_JUMP_THUNK_PARAM;
+#elif defined(_TARGET_ARM64_)
+            // Prevent indir->gtRegNum from colliding with addr->gtRegNum
+            indir->gtRegNum = REG_JUMP_THUNK_PARAM;
+
+            // Sanity checks
+            assert(addr->gtRegNum != indir->gtRegNum); // indir and addr registers must be different
```

Comment:
I think this would be clearer if you kept the assignment to `indir->gtRegNum` as common code, and then added the asserts for ARM64 only.
I don't really know what the special cases are about, e.g. why the check below for `CORINFO_CORERT_ABI`, but as @BruceForstall pointed out, this needs general cleanup, so I'm not going to press for a minor clarification.